### DISCUSSION TITLE
Does aiobotocore allow you to pre-initialize clients to deal with Lambda cold starts?

### DISCUSSION THREAD

**Discussion by tibbe** [aio-libs/aiobotocore#987] [Q&A]
We're currently using regular `boto3` and to avoid Lambda cold start overhead we are creating sessions/clients at import time and store them in global variables. Something like:

```py
# Simplified from actual code:
dynamodb_resource = boto3.Session().resource('dynamodb')
```

(We also use Lambda provisioned concurrency so not even the first request sees cold start overheads.)

What's the equivalent in `aiobotocore`? The examples seem to create the client just before making the request, which presumably incurs the cost of client creation.

(If it matter we use FastAPI to create our server. Perhaps we should try to hold some state in our server object.)

**Comment by thehesiod** [comment]
see https://github.com/terrycain/aioboto3 for boto3 equivalents

**Comment by tibbe** [comment]
Note that I'm not looking for boto3 APIs. I'm looking for caching of async clients. The aioboto3 examples do the same thing I described above, namely initialize clients just before using them.

**Reply by thehesiod** [reply]
You can cache clients they same way. I'm not sure what your question is

**Reply by tibbe** [reply]
Maybe I just don't know how async Python works. I just assumed I couldn't run async actions at the top-level at import time.

**Reply by thehesiod** [reply]
You can but it's not good practice

**Reply by thehesiod** [reply]
For our lambdas i store a ref as a global in a singleton

**Reply by tibbe** [reply]
At risk of sounding very naive, so in your lambdas you do call async functions at import time, since the create client functions in this library are async, and store them in a global singleton?

**Reply by flpStrri** [reply]
I can be wrong but you can call the async client init inside the handler and make the client creation only once (running it only if the global variable outside the handler is `None`). You can "hide it" in a separate module that outside of it seems you are simply asking for the handler with an async func.

**Reply by tibbe** [reply]
@flpStrri this works as a way to initialize only once but unfortunately it loses the value of Lambda provisioned concurrency, which is to initialize the Lambda *before the first request*, to avoid the first request seeing increased latency. Basically initialization has to happen at import time for this to work.

**Reply by flpStrri** [reply]
True! So probably the way to do it is:
- create a event loop;
- run the init code;
- close the event loop;
- let the handler code do its magic.

About that, I remember handling some libs using aiohttp and the init phase of the context manager was actually sync. The context manager was async because of the teardown method. This is not the case here, right?

**Reply by thehesiod** [reply]
here's what we do in our lambda function that has been running for years
```python
class Globals:
    def __init__(self):
        self.logger = logging.getLogger(...)
        self._exit_stack = AsyncExitStack()
        self.logger.info("Starting Globals init")

        self.boto_config = botocore.config.Config(connect_timeout=TIMEOUT_S, read_timeout=TIMEOUT_S)
        self.boto_session: Optional[aiobotocore.AioSession] = None
        self.s3_client = None
        self.kms_client = None
        self.dynamo_client = None
        self.dynamo_counter: Optional[DynamoCounter] = None
        self.logger.info("Finished Globals init")

    async def __aenter__(self):
        self.boto_session = aiobotocore.session.get_session()

        self.s3_client = await self._exit_stack.enter_async_context(self.boto_session.create_client('s3', config=self.boto_config))
        self.kms_client = await self._exit_stack.enter_async_context(self.boto_session.create_client('kms', config=self.boto_config))
        self.dynamo_client = await self._exit_stack.enter_async_context(self.boto_session.create_client('dynamodb', config=self.boto_config))
        self.dynamo_counter = DynamoCounter(self.dynamo_client, self.logger, DYNAMO_STATS_TABLE_NAME, DYNAMO_STATS_KEY_NAME)

        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self._exit_stack.__aexit__(exc_type, exc_val, exc_tb)


GLOBALS: Optional[Globals] = None

...

GLOBALS = GLOBALS or asyncio.get_event_loop().run_until_complete(Globals(AppNames.LABELER_APP_NAME, SCOPES).__aenter__())
```

**Reply by thehesiod** [reply]
remember each thread can only have one run loop, so you don't need to worry about multiple run loops

**Reply by flpStrri** [reply]
@thehesiod, will the __aexit__ be called? I mean in the lambda runtime. I know that if you use the `Globals` as an async context manager on pytest it exit the context and clean the whole "context"... but not in lambda right?

**Reply by tibbe** [reply]
Related question. Does it have to be an async context in this situation? It seems like you call `__aenter__` manually.

**Reply by flpStrri** [reply]
@tibbe thats basically what I asked but more clearly. But I think that the context managers is used to be easier to handle globals on tests. I would do it. 😄

**Reply by thehesiod** [reply]
It's mostly there for integration tests to avoid unclosed sockets. What I've found is that aws can terminate your instance at any moment so any sort of shutdown is just for testing.

**Comment by vasily-kartashov** [comment]
@tibbe I was wondering if you've got anywhere with this one? I am rather surprised there's no information on this question. For example timestream client takes ages to initialize, so keeping the client between "warm" invocations would be obviously beneficial. After all the only reason we use aiobotocore is performance.

**Reply by thehesiod** [reply]
this is more of a botocore question

**Reply by vasily-kartashov** [reply]
With boto3 I can just create a client in a module outside the handler, et voila:
```
client = boto3.client('...')

def handle(event, context):
    ...
```

I ended up making default loop waiting for
```
client = wait session.create_client().__aenter__()
```

Which is not pretty to say the least.

**Comment by thehesiod** [comment]
seems like I answered this sufficiently, let me know otherwise :)
